internal/services/newrelic/new_relic_tag_rule_resource.go

```diff
@@ -0,0 +1,413 @@
+}
+
+type FilteringTagModel struct {
+	Action tagrules.TagAction `tfschema:"action"`
+	Name   string             `tfschema:"name"`
+	Value  string             `tfschema:"value"`
+}
+
+type NewRelicTagRuleResource struct{}
+
+var _ sdk.ResourceWithUpdate = NewRelicTagRuleResource{}
+
+func (r NewRelicTagRuleResource) ResourceType() string {
+	return "azurerm_new_relic_tag_rule"
+}
+
+func (r NewRelicTagRuleResource) ModelObject() interface{} {
+	return &NewRelicTagRuleModel{}
+}
+
+func (r NewRelicTagRuleResource) IDValidationFunc() pluginsdk.SchemaValidateFunc {
+	return tagrules.ValidateTagRuleID
+}
+
+func (r NewRelicTagRuleResource) Arguments() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{
+		"monitor_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: monitors.ValidateMonitorID,
+		},
+
+		"aad_log_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+
+		"activity_log_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+
+		"log_tag_filter": r.tagFilterSchema(),
+
+		"metric_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+
+		"metric_tag_filter": r.tagFilterSchema(),
+
+		"subscription_log_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+}
+
+func (r NewRelicTagRuleResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+func (r NewRelicTagRuleResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			var model NewRelicTagRuleModel
+			if err := metadata.Decode(&model); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			client := metadata.Client.NewRelic.TagRulesClient
+			monitorId, err := monitors.ParseMonitorID(model.NewRelicMonitorId)
+			if err != nil {
+				return err
+			}
+
+			id := tagrules.NewTagRuleID(monitorId.SubscriptionId, monitorId.ResourceGroupName, monitorId.MonitorName, "default")
+			existing, err := client.Get(ctx, id)
+			if err != nil && !response.WasNotFound(existing.HttpResponse) {
+				return fmt.Errorf("checking for existing %s: %+v", id, err)
+			}
+
+			if !response.WasNotFound(existing.HttpResponse) {
+				return metadata.ResourceRequiresImport(r.ResourceType(), id)
+			}
+
+			logRules := tagrules.LogRules{
+				FilteringTags:        expandFilteringTagModelArray(model.LogTagFilter),
+				SendAadLogs:          pointer.To(tagrules.SendAadLogsStatusDisabled),
+				SendActivityLogs:     pointer.To(tagrules.SendActivityLogsStatusDisabled),
+				SendSubscriptionLogs: pointer.To(tagrules.SendSubscriptionLogsStatusDisabled),
+			}
+
+			email, err := r.getEmail(ctx, metadata, monitorId)
+			if err != nil {
+				return err
+			}
```

Comment:
`getEmail` is not separate. It's used in both `create` and `update`

Reply:
It would be good to get visibility on what a function is doing, i.e. what API it's interacting with - so instead of passing in metadata could we pass in the specific clients that are needed instead?

Reply:
code updated. Thanks.